code/go/pkg/validator/validator_test.go

```diff
@@ -119,6 +119,17 @@ func TestValidateFile(t *testing.T) {
 				fmt.Sprintf("field ilm_policy: ILM policy \"logs-bad_custom_ilm_policy.test-notexists\" not found in package, expected definition in \"%sbad_custom_ilm_policy/data_stream/test/elasticsearch/ilm/notexists.json\"", osTestBasePath),
 			},
 		},
+		"bad_select": {
+			"data_stream/foo_stream/manifest.yml",
+			[]string{
+				"field streams.0.vars.1: Must validate \"then\" as \"if\" was valid",
+				"field streams.0.vars.1: options is required",
+				"field streams.0.vars.2: Must validate \"then\" as \"if\" was valid",
+				"field streams.0.vars.2.options: Invalid type. Expected: array, given: null",
+				"field streams.0.vars.3: Must validate \"else\" as \"if\" was not valid",
+				"field streams.0.vars.3: Must not validate the schema (not)",
```

Comment:
Do these error end up in the user output? They look weird.

Reply:
I agree they are not very helpful. I guess they will show up in the user output - is there a way to test this now? At the moment `elastic-package check` doesn't know about select and options.

Reply:
You can build `elastic-package` replacing the package-spec dependency.
For that you can run these commands:
```
go mod edit -replace github.com/elastic/elastic-package2=github.com/jillguyonnet/elastic-package@<your_latest_changeset_in_PR>

go mod tidy 

make build
```

Reply:
Thanks! Yep, confirmed, this is the output. As you said, I guess we can't do much about it.
```
Error: checking package failed: linting package failed: found 6 validation errors:
   1. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.1: Must validate "then" as "if" was valid
   2. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.1: options is required
   3. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.2: Must validate "then" as "if" was valid
   4. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.2.options: Invalid type. Expected: array, given: null
   5. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.3: Must validate "else" as "if" was not valid
   6. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.3: Must not validate the schema (not)
```

Reply:
We could filter-out certain errors. Here I think it'd be safe to assume that we can remove all these errors starting with `Must validate`/`Must not validate`. This could be done before returning the errors in `ValidatePackage`.

This would leave these errors, that are still a bit redundant, but more meaningful for users:
```
Error: checking package failed: linting package failed: found 2 validation errors:
   1. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.1: options is required
   2. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.2.options: Invalid type. Expected: array, given: null
```

Reply:
For these two error messages, it should be shown something to the user 
```
   5. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.3: Must validate "else" as "if" was not valid
   6. file "/Users/jillguyonnet/Code/jillguyonnet/package-spec/test/packages/bad_select/data_stream/foo_stream/manifest.yml" is invalid: field streams.0.vars.3: Must not validate the schema (not)
```

Those errors are for the case that options is defined in a different type (e.g. string). However, those messages do not show helpful information about it :/

Reply:
I think it would be OK to only have 1 error instead of 2, so 3 instead of 6 in the example above:
```
options is required
Invalid type. Expected: array, given: null
Must not validate the schema (not)
```
As you said, however, the third one is really unhelpful, it really should be something like `options is forbidden when type is not select` or something similar. I guess in addition to filtering errors we could transform the message... but I'm not sure it can be done in a way that will scale to future conditional cases.
The second error is also confusing (the issue is that the array is empty), but it's a bit less bad.

Reply:
I've been trying to set up a custom error message in the JSON schema without luck so far. I can add filtering if you want on the following errors:
```
"Must validate \"then\" as \"if\" was valid"
"Must validate \"else\" as \"if\" was not valid"
```
but I think we should keep `Must not validate the schema (not)`, otherwise we'll silence the error when options should not be present, which is worse than a bad error message. :)
Let me know your thoughts.

Reply:
Ok. Maybe we can replace `Must not validate the schema (not)` with `Must not be present`, around the same place where we do the filtering for the other errors.

Reply:
Ok, I've pushed a tentative solution (Mario warned me I'd be missing map and filter on arrays, he was right 😆), let me know of any improvements.
Note that there already was an error message with `Must not validate the schema (not)` [elsewhere](https://github.com/elastic/package-spec/pull/486/files#diff-4dd1d701e9e3346e2b2ccdf83fbb4f2f94f3ae3373a8cfe7c962716fc3dc323bR101), I modified it too.

Reply:
> I'd be missing map and filter on arrays

Now that there are generics in Go there are some attempts :smile: https://pkg.go.dev/golang.org/x/exp/slices